src/States.cpp

```diff
@@ -0,0 +1,291 @@
+    // extract positions from state vector
+    std::vector<ale::Vec3d> positions;
+
+    for(ale::State state : m_states) {
+        positions.push_back(state.position);
+    }
+    return positions;
+  }
+
+
+  std::vector<ale::Vec3d> States::getVelocities() const {
+    // extract velocities from state vector
+    std::vector<ale::Vec3d> velocities;
+
+    for(ale::State state : m_states) {
+        velocities.push_back(state.velocity);
+    }
+    return velocities;
+  }
+
+
+  std::vector<double> States::getTimes() const {
+    return m_ephemTimes;
+  }
+
+
+  int States::getReferenceFrame() const {
+    return m_refFrame;
+  }
+
+
+  bool States::hasVelocity() const {
+    std::vector<ale::Vec3d> velocities = getVelocities();
+    bool allZero = std::all_of(velocities.begin(), velocities.end(), [](ale::Vec3d vec)
+                               { return vec.x==0.0 && vec.y==0.0 && vec.z==0.0; });
+    return !allZero;
+  }
+
+
+  ale::State States::getState(double time, ale::interpolation interp) const {
+    std::vector<double> xs, ys, zs, vxs, vys, vzs;
+    int lowerBound = ale::interpolationIndex(m_ephemTimes, time);
+    std::vector<double> interpTimes = {};
+    int interpStart;
+    int interpStop;
+
+    if (m_ephemTimes.size() <= 3) {
+      interpStart = 0;
+      interpStop = m_ephemTimes.size() - 1;
+    }
+    else if (lowerBound == 0) {
+      interpStart = lowerBound;
+      interpStop = lowerBound + 3;
+    }
+    else if (lowerBound == m_ephemTimes.size() - 1) {
+      interpStart = lowerBound - 3;
+      interpStop = lowerBound;
+    }
+    else if (lowerBound == m_ephemTimes.size() - 2) {
+      interpStart = lowerBound - 2;
+      interpStop = lowerBound + 1;
+    }
+    else {
+      interpStart = lowerBound - 1;
+      interpStop = lowerBound + 2;
+    }
+
+    ale::State state;
+
+    for (int i = interpStart; i <= interpStop; i++) {
+      state = m_states[i];
+      interpTimes.push_back(m_ephemTimes[i]);
+      xs.push_back(state.position.x);
+      ys.push_back(state.position.y);
+      zs.push_back(state.position.z);
+      vxs.push_back(state.velocity.x);
+      vys.push_back(state.velocity.y);
+      vzs.push_back(state.velocity.z);
+    }
+
+    ale::Vec3d position, velocity;
+
+    if ( interp == LINEAR || (interp == SPLINE && !hasVelocity())) {
+      position = {interpolateState(xs,  interpTimes, time, interp, 0),
+                  interpolateState(ys,  interpTimes, time, interp, 0),
+                  interpolateState(zs,  interpTimes, time, interp, 0)};
+
+      velocity = {interpolateState(vxs, interpTimes, time, interp, 0),
+                  interpolateState(vys, interpTimes, time, interp, 0),
+                  interpolateState(vzs, interpTimes, time, interp, 0)};
+    }
+    else if (interp == SPLINE && hasVelocity()){
+      // Do hermite spline if velocities are available
+      double baseTime = (m_ephemTimes.at(0) + m_ephemTimes.at(m_ephemTimes.size() - 1)) / 2.;
```

Comment:
I would change this to `(interpTimes.front() + interpTimes.back()) / 2`. It won't change the output of the function, but it will make it a little more numerically stable by interpolating closer to 0.